incubator/hnc/pkg/reconcilers/hierarchical_namespace.go

```diff
@@ -31,44 +39,176 @@ type HierarchicalNamespaceReconciler struct {
+		err := r.writeInstance(ctx, log, inst)
+		return ctrl.Result{}, err
+	}
+
+	// Get the self-serve namespace's hierarchyConfig and namespace instances.
+	hcInst, nsInst, err := r.hcr.GetInstances(ctx, log, nm)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	r.syncWithForest(log, inst, nsInst, hcInst)
+
+	// TODO report the "RequiredChildConflict" in the HNS reconciliation. See issue:
+	//  https://github.com/kubernetes-sigs/multi-tenancy/issues/490
+
+	err = r.writeInstance(ctx, log, inst)
+	return ctrl.Result{}, err
+}
+
+// HierarchicalNamespace (HNS) is synced with the in-memory forest, the HierarchyConfig and
+// the namespace instances to update its HNS state. This will be the only place the "Owner"
+// field in the forest is set. Therefore, the "Owner" field can be used in the forest to get
+// all the HNS objects of a namespace.
+func (r *HierarchicalNamespaceReconciler) syncWithForest(log logr.Logger, inst *api.HierarchicalNamespace, nsInst *corev1.Namespace, hcInst *api.HierarchyConfiguration) {
+	r.forest.Lock()
+	defer r.forest.Unlock()
+
+	// Names of the hierarchical namespace and the current namespace.
+	nm := inst.Name
+	pnm := inst.Namespace
+
+	// Get the namespace instance in memory.
+	ns := r.forest.Get(nm)
+
+	// If the real namespace instance doesn't exist yet, update forest and enqueue the namespace
+	// to HierarchyConfig reconciler.
+	if nsInst.Name == "" {
+		log.Info("The self-serve subnamespace does not exist", "namespace", nm)
 
-	// Enqueue the in-momery hierarchyConfig instance of the hierarchical namespace.
-	// The hierarchyConfig reconciler will create the namespace and hierarchyConfig
-	// instances on apiserver accordingly.
-	reason := "new/updated hierarchical namespace"
-	r.hcr.enqueueAffected(log, reason, nm)
+		// Set the "Owner" in the forest of the hierarchical namespace to the current namespace.
+		log.Info("Setting the subnamespace's owner in the forest", "owner", pnm, "namespace", nm)
+		// TODO rename RequiredChildOf to Owner in the forest. See issue:
+		//  https://github.com/kubernetes-sigs/multi-tenancy/issues/469
+		ns.RequiredChildOf = pnm
 
-	// TODO sync with forest to report conflicts in hns.Status.State. See issue:
-	//  https://github.com/kubernetes-sigs/multi-tenancy/issues/487
+		// Enqueue the not-yet existent self-serve subnamespace. The HierarchyConfig reconciler will
+		// create the namespace and the HierarchyConfig instances on apiserver accordingly.
+		reason := "new/updated hierarchical namespace"
+		r.hcr.enqueueAffected(log, reason, nm)
 
-	return ctrl.Result{}, nil
+		inst.Status.State = api.Missing
+		return
+	}
+
+	log.Info("The self-serve subnamespace already exists", "namespace", nsInst.Name)
+	// If the namespace is not a self-serve namespace or it's a self-serve namespace of another
+	// namespace. Report the conflict.
+	if nsInst.Annotations[api.AnnotationOwner] != pnm {
+		log.Info("The namespace cannot be used as a self-serve subnamespace")
```

Comment:
Maybe be more specific here? Include a log about what the conflict was (eg the value of the annotation).